File: backend/src/dto/graph.rs
use serde::{Deserialize, Serialize};

/// Node types in the legal proof graph hierarchy
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum GraphNodeType {
    LegalCount,
    Allegation,
    MotionClaim,
    Evidence,
    Document,
}

/// A node in the graph visualization
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphNode {
    pub id: String,
    pub label: String,
    pub node_type: GraphNodeType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subtitle: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
}

/// An edge connecting two nodes
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct GraphEdge {
    pub source: String,
    pub target: String,
    pub relationship: String,
}

/// Response containing graph data for visualization
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphResponse {
    pub nodes: Vec<GraphNode>,
    pub edges: Vec<GraphEdge>,
    pub title: String,
    pub hierarchy_type: String,
}

File: frontend/src/services/graph.ts
import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";

/** Node type from the graph API. Now a string to support any schema-defined type. */
export type GraphNodeType = string;

export type GraphNode = {
  id: string;
  label: string;
  node_type: GraphNodeType;
  subtitle?: string;
  details?: string;
};

export type GraphEdge = {
  source: string;
  target: string;
  relationship: string;
};

export type GraphResponse = {
  nodes: GraphNode[];
  edges: GraphEdge[];
  title: string;
  hierarchy_type: string;
};

export async function getLegalProofGraph(
  countId?: string
): Promise<GraphResponse> {
  const url = countId
    ? `${API_BASE_URL}/api/graph/legal-proof?count_id=${encodeURIComponent(countId)}`
    : `${API_BASE_URL}/api/graph/legal-proof`;

  const response = await authFetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch graph: ${response.status}`);
  }

  return response.json();
}
